swarm/src/listen_opts.rs

```diff
@@ -0,0 +1,33 @@
+use crate::ListenerId;
+use libp2p_core::Multiaddr;
+
+#[derive(Debug)]
+pub struct ListenOpts {
+    id: ListenerId,
+    address: Multiaddr,
+}
+
+impl ListenOpts {
+    pub fn new(address: Multiaddr) -> ListenOpts {
+        ListenOpts {
+            id: ListenerId::next(),
+            address,
+        }
+    }
+
+    /// Get the [`ListenerId`] of this listen attempt
+    pub fn listener_id(&self) -> ListenerId {
+        self.id
+    }
+
+    /// Get the [`Multiaddr`] that is being listened on
+    pub fn address(&self) -> Multiaddr {
+        self.address.clone()
+    }
```

Comment:
```suggestion
    pub fn address(&self) -> &Multiaddr {
        &self.address
    }
```

The user can always clone themselves but never undo the clone.